include/nbl/video/IGPUDescriptorSet.h

```diff
@@ -28,10 +30,131 @@ class NBL_API IGPUDescriptorSet : public asset::IDescriptorSet<const IGPUDescrip
 		using base_t = asset::IDescriptorSet<const IGPUDescriptorSetLayout>;
 
 	public:
-		IGPUDescriptorSet(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& _layout) : base_t(std::move(_layout)), IBackendObject(std::move(dev)) {}
+		IGPUDescriptorSet(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& _layout, core::smart_refctd_ptr<IDescriptorPool>&& pool)
+			: base_t(std::move(_layout)), IBackendObject(std::move(dev)), m_version(0ull), m_pool(std::move(pool))
+		{
+            allocateDescriptors();
+
+            for (auto i = 0u; i < asset::EDT_COUNT; ++i)
+            {
+                // There is no descriptor of such type in the set.
+                if (m_descriptorStorageOffsets.data[i] == ~0u)
+                    continue;
+
+                const auto type = static_cast<asset::E_DESCRIPTOR_TYPE>(i);
+
+                // Default-construct the core::smart_refctd_ptr<IDescriptor>s because even if the user didn't update the descriptor set with ILogicalDevice::updateDescriptorSet we
+                // won't have uninitialized memory and destruction wouldn't crash in ~IGPUDescriptorSet.
+                std::uninitialized_default_construct_n(getDescriptorStorage(type) + m_descriptorStorageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
+            }
+
+            const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
+            if (mutableSamplerCount > 0)
+                std::uninitialized_default_construct_n(getMutableSamplerStorage() + m_descriptorStorageOffsets.data[asset::EDT_COUNT], mutableSamplerCount);
+        }
+
+        inline uint64_t getVersion() const { return m_version.load(); }
+
+        inline void incrementVersion() { m_version.fetch_add(1ull); }
```

Comment:
this should not be public, declare friendship with logical device

Reply:
`incrementVersion` that is

Reply:
Resolved.